Python: Type Hints, Misc Cleanup, Drop Python 3.6/3.7 Support

Cuphat opened the pull request:
Drops support for Python 3.6/3.7 and adds type hints to all class/instance attributes and function definitions to make it easier to know what variables contain while working on the code.

These are likely far from perfect as I learned while working through all of them, but at the very least nothing should be broken.

Comparing the spoiler outputs from running tests on Dev and this branch yields a few differences:

- When `hint_dist_user` is modified by `World`, those changes no longer modify it in the original settings that get output to the spoiler log as setting values are now deep copied. This seems like better behavior in general, though it does make me question why `hint_dist_user` is modified in `World` to begin with.
- The spoilers from the preset tests have slightly different ordering to the starting items. Whatever.
- Playthroughs in some of the tests with entrance rando appear to be slightly different for reasons I don't feel like debugging right now. The actual seed is the same, though I'll probably look into it at some point.
- A few tests had their plandos updated, as I renamed the GUI's `starting_items` setting to `starting_inventory` so that we don't have to try to accept two completely different formats into the same setting name. Those obviously have completely different output, as the plando file is part of the seed.

Cuphat committed:
Purge the useless `window` that hasn't done anything in 3 years.

Cuphat committed:
Misc Cleanup

Removed `random_choices` shim as Python 3.5 support is not coming back.
Simplified the Python version checks on `OoTRandomizer.py` and `Gui.py`.
Lots of minor code-style fixes.

Cuphat committed:
Type Annotations and Cleanup

Added type hints for `crc.py`, MBSDIFFPatch.py`, `ntype.py`, and `Rom.py`.
Reworked the "Record" types and initialized the instance attributes in `Plandomizer.py` to make it more clear what attributes exist and where.

Cuphat committed:
SettingsList Revamp

Remake SettingInfos as a class with descriptors instead of a simple dictionary.

Cuphat committed:
Type hints and other cleanup.

Adds type hints to all class/instance attributes and function definitions to make it easier to know what variables contain while working on the code.

These are likely far from perfect as I learned while working through all of them, but at the very least nothing should be broken.

fenhl commented:
Could you add some docs explaining how to run the typechecker?

fenhl commented:
Overall this PR is a huge improvement in code style, but I don't agree with the excessive indentation added to some multiline statements. It makes things harder to read because you need extra horizontal scrolling, and harder to format because lining things up based on the syntax in the previous line means you can't just use tab or shift+tab to indent.

fenhl commented:
You may want to run it in VS Code as well, since many contributors use that editor. The extension that provides Python type checking is called Pylance, and its type checker is off by default. Even then, it reports 6 problems all related to type annotations on this branch. If I enable basic type checking, the report count goes above 1000.

Cuphat commented:
Thanks for the tip to check Pylance. There are definitely some botches in there which are easily fixed. 114 "errors" from it in Sounds.py are absolutely psychotic and not actually errors in the code but instead it not understanding the concept of conditionally using dataclass or namedtuple depending on Python version (also another reason to drop Python 3.6 support, but that's a different potential conversation). And a lot of them are just assumptions baked into the code that get a spotlight shined on them.

I'm gonna work through and fix what I can.

fenhl commented:
We should probably just drop Python 3.6. It's been eol for over a year and even Debian oldstable is on 3.7 already. And being able to use dataclasses everywhere would allow us to do a bunch of cleanup, and `from __future__ import annotations` lets you consistently leave off the quotes around type annotations.

fenhl commented:
There's also [`mypy`](https://www.mypy-lang.org/), a CLI type checker which just hit 1.0. We could eventually integrate it into our CI.

Cuphat committed:
Type Fixes, More Cleanup

Cuphat committed:
Drop Python 3.6 Support

Main features of Python 3.7 being utilized is `from __future__ import annotations` and dataclasses.

Cuphat commented:
There are still a ton of errors when type checking is enforced, but those are mainly design flaws in rando itself that I am considering out of scope for this particular PR. My goal will be to fix all of them, but I don't want to hold off getting this merged in the meantime.

Cuphat committed:
Rework how Dungeon item collections are populated.

Cuphat committed:
Reduce Iteration

Cuphat committed:
Drop Python 3.7 Support

This lets us get rid of some wacky conditionals and allows us to use the walrus operator where appropriate.

Cuphat committed:
Rework World/etc copy to utilize a dictionary of objects copied.

Cuphat committed:
Rework `get_unrestricted_dungeon_items`, etc.

This changes seeds due to the ordering of dungeon items from these functions differing from the previous version.

Cuphat commented:
Now that this is rebased, things left for me to do:

Confirm that it still actually works outside of the unittests.
Confirm that the differences in spoiler output are still trivial or at least not bugs.
Look through and make sure that everything is type hinted, including things that were added between 7.1.70 and now.
Other dumb shit I am sure.

Cuphat committed:
Type hints for new methods.

todo: squash most of the later commits

Cuphat committed:
DMA `free_space` revamp.

cjohnson57 commented on Dungeon.py:
```diff
@@ -1,80 +1,115 @@
-import os
+from __future__ import annotations
```
For these `__future__` imports, I googled and found that this basically means it's using features from future python versions. Once these features are part of the next main python versions, what will happen here? Will those things be removed from the `__future__` library, meaning we'll have to start requiring the newest Python version which presumably includes `annotations`?

Cuphat replied:
`__future__` imports are special as they don't actually import anything, rather they set a flag saying "enable this feature". Having the `__future__` import on a version where that feature is now default just doesn't do anything. From the python docs:

> No feature description will ever be deleted from [__future__](https://docs.python.org/3/library/__future__.html#module-__future__).

cjohnson57 replied:
Interesting, sounds good then

cjohnson57 commented on Dungeon.py:
```diff
@@ -1,80 +1,115 @@
                 region.dungeon = self
                 self.regions.append(region)
 
+    def copy(self, *, copy_dict: Optional[dict[int, Any]] = None) -> Dungeon:
```
Can you explain these changes more? I don't really understand them. Is new_world replaced by copy_dict, which is now an optional parameter for some reason? And what does the * mean?

Cuphat replied:
`*` just means that any parameter to the right of it has to be a named parameter and can't be used as a positional parameter.

Basically, the change is that these copy functions will now make copies of all the shit attached to them. The `copy_dict` is where references to those new copies are stored. It is using the python `id` of the old version of the object as the key so each thing is only copied once. It's optional because if it is not provided, one is created inside of the first copy method call before being passed down to all the resulting copy calls. The result should be that the copies created at the end have references to the copies of all the things they referenced before instead of doing some of the funky shit that we were doing to make it work.

cjohnson57 replied:
It's still a little bit over my head but I think I understand well enough

cjohnson57 commented on Fill.py:
```diff
@@ -534,18 +531,14 @@ def fill_restrictive_fast(window, worlds, locations, itempool):
         spot_to_fill = locations.pop()
         item_to_place = itempool.pop()
         spot_to_fill.world.push_item(spot_to_fill, item_to_place)
-        window.fillcount += 1
```
Where is the functionality for updating the GUI window replaced since it was removed from fill?

Cuphat replied:
That functionality only existed for the tkinter GUI. The node GUI has always just read the CLI output.

cjohnson57 replied:
oh lmao, I didn't realize it was that much of a relic...

cjohnson57 commented on Hints.py:
```diff
@@ -1,29 +1,50 @@
-    "Bottle", "Bottle with Red Potion","Bottle with Green Potion", "Bottle with Blue Potion",
+from Utils import data_path
+
+if sys.version_info >= (3, 10):
+    from typing import TypeAlias
+else:
+    TypeAlias = str
```
Feels a bit awkward having this in all these different files, I wouldn't be too opposed to just requiring 3.10 at some point (Depending on the answer to the `__future__` question might have to require a newer version again anyway?)

Cuphat replied:
Yeah, it's a bit ugly but ¯\_(ツ)_/¯

cjohnson57 replied:
I mean, honestly, we're already bumping the python requirement up, is there any reason to not just go all the way and bump it up to 3.10? (Although that might be more of a question for TreZ)

Cuphat replied:
I think when we discussed it in Discord, we eventually settled on not dropping support for 3.9 as it is still a very common version out in the wild. I never went and dropped 3.8 support because I am not sure I saw anything new to 3.9 being particularly relevant.

cjohnson57 commented on Item.py:
```diff
@@ -49,103 +55,87 @@ def __init__(self, name='', event=False):
-
-        self.solver_id = self.info.solver_id
+        self.info: ItemInfo = ItemInfo.events[name] if event else ItemInfo.items[name]
+        self.price: Optional[int] = self.info.special.get('price', None)
```
What does `Optional` mean when it's not a function parameter? Just a way of saying nullable basically? (Even though I think technically all types in Python are nullable lol)

Cuphat replied:
Optional just means that it can be None, yeah.

cjohnson57 commented on ItemPool.py:
```diff
@@ -407,24 +409,25 @@ def generate_itempool(world):
-    world.available_tokens = placed_items_count.get("Gold Skulltula Token", 0) \
-                           + pool.count("Gold Skulltula Token") \
-                           + world.distribution.get_starting_item("Gold Skulltula Token")
+    world.available_tokens = (placed_items_count.get("Gold Skulltula Token", 0)
+                              + pool.count("Gold Skulltula Token")
+                              + world.distribution.get_starting_item("Gold Skulltula Token"))
```
Is it standard styling to include the surrounding parenthesis in this case?

Cuphat replied:
Surrounding parenthesis are required to not have to escape newlines.

cjohnson57 replied:
Huh, didn't know about that interaction

cjohnson57 commented on Location.py:
```diff
@@ -71,63 +88,89 @@ def add_rule(self, lambda_rule):
-def LocationFactory(locations, world=None):
+@overload
+def LocationFactory(locations: str) -> Location:
+    pass
+
+
+@overload
+def LocationFactory(locations: list[str]) -> list[Location]:
+    pass
+
+
+def LocationFactory(locations: str | list[str]) -> Location | list[Location]:
```
Can you explain this?

Cuphat replied:
Since python doesn't have real overloaded functions, the `@overload` versions of it above just serve as type hints. In this case, it lets your IDE/whatever type parser know that if you send it a single string, you get back a single location, and if you pass it a list of strings, you get back a list of locations. Without the `@overload`s, each return is assumed to be either a location or a list of locations - it doesn't know which one would have been returned.

cjohnson57 replied:
Huh, alright

cjohnson57 commented on Plandomizer.py:
```diff
@@ -44,112 +52,120 @@ class InvalidFileException(Exception):
-            return dump_obj(self.to_json())
-    return Record
+class DungeonRecord(Record):
+    mapping: dict[str, Optional[bool]] = {
```
Nitpick, maybe rename this to `mq_mapping`?

Cuphat replied:
That'd probably make sense. My process of going through `Plandomizer.py` was more focused on actually defining the properties that it was currently using so it was a bit easier to work with and less on renaming those properties.

cjohnson57 commented on SettingTypes.py:
```diff
@@ -0,0 +1,345 @@
+        else:
+            self.bitwidth = 0
+
+        # default value if undefined/unset
+        self.default = default
+        if self.default is None:
+            if self.type == bool:
+                self.default = False
+            elif self.type == str:
+                self.default = ""
+            elif self.type == int:
+                self.default = 0
+            elif self.type == list:
+                self.default = []
+            elif self.type == dict:
+                self.default = {}
```
Could this not be handled in the init for the `SettingInfoX` classes?

Cuphat replied:
Possibly. That's mainly just a relic of the old code that wasn't fully updated because it still works out alright like that.

cjohnson57 commented on Models.py:
```diff
@@ -54,30 +61,30 @@ def SetBase(self, base):
             self.base = RUNNING_MAN_START
 
-    def GoTo(self, dest):
+    def GoTo(self, dest: int) -> None:
```
Is there a particular reason you didn't give this file the treatment of converting all the camel/pascal case into snake case?

Cuphat replied:
Mainly just that it is a pretty self-contained thing so I didn't spend as much time on that file.

cjohnson57 commented on World.py:
```diff
@@ -1,99 +1,101 @@
-                or settings.warp_songs or settings.spawn_positions
-            )
-        ):
+        self.ensure_tod_access: bool = bool(self.shuffle_interior_entrances or settings.shuffle_overworld_entrances or settings.spawn_positions)
```
Why is specifying bool on the right side of the assignment done here?

Cuphat replied:
That's where that attribute is first defined and I tried to add a type hint for all class/instance attributes when they are defined. `World` just has a sprawling `__init__` so it looks a bit weirder.

Cuphat replied:
Also, I had to wrap the thing in a `bool()` call so that the result of it was actually a bool since two of those settings it uses are actually a list and calling `or` just gets the first result that would be considered True, it doesn't itself return a bool.

cjohnson57 replied:
Should the same thing not then be done for the ones below this like `disable_trade_revert`?

Edit: NVM didn't see your second comment before I commented

Cuphat replied:
It probably should be done for `disable_trade_revert` for the same reason. I'm not sure why I only did it for `ensure_tod_access`, I think my IDE just happened to bitch about that one at some point?

Cuphat replied:
Just looked at this harder and `disable_trade_revert` is an or of 3 booleans which is why it didn't need the `bool()` call. `ensure_tod_access` is an or of 2 booleans and a list, so it does.

cjohnson57 commented on World.py:
```diff
@@ -1009,107 +1034,59 @@ def set_goals(self):
-            for region in self.regions:
-                for r_location in region.locations:
-                    if r_location.name == location:
-                        self._location_cache[location] = r_location
-                        return r_location
-        raise KeyError('No such location %s' % location)
-
-
-    def get_items(self):
-        return [loc.item for loc in self.get_filled_locations()] + self.itempool
 
+        for region in self.regions:
+            for location in region.locations:
+                self._location_cache[location.name] = location
+        if (location := self._location_cache.get(location_name, None)) is not None:
+            return location
+        raise KeyError('No such location %s' % location_name)
 
-    def get_itempool_with_dungeon_items(self):
-        return self.get_restricted_dungeon_items() + self.itempool
-
+    def get_itempool_with_dungeon_items(self) -> Iterator[Item]:
+        yield from self.get_restricted_dungeon_items()
+        yield from self.itempool
```
IIs this change purely stylistic or is there an advantage of doing it this way?

Cuphat replied:
The previous ones combines two lists and returns a new list, whereas `yield from` doesn't create a new list, just allows iteration through the existing lists. Since we only iterate over the result of that function call, there's no reason to always make an intermediary list.

cjohnson57 replied:
Fair enough

cjohnson57 reviewed (commented):
Amazing work, I certainly wouldn't have had the patience to do this lol

cjohnson57 commented:
> Overall this PR is a huge improvement in code style, but I don't agree with the excessive indentation added to some multiline statements. It makes things harder to read because you need extra horizontal scrolling, and harder to format because lining things up based on the syntax in the previous line means you can't just use tab or shift+tab to indent.

Unless this has been resolved, can you point some of these out?

As far as I can tell, everything here follows the standard for this, that standard being to line the subsequent lines up with the start of the statement on the original line (unless it's an if statement in which case you add more tabs so that it doesn't get confused for the code contained within the if). And viewing everything on Github, nothing got too close to the edge of my screen, and I just have a standard 1080p monitor

Cuphat committed:
Fix bugs found while doing web testing.

Settings now return the default value if they are set to null for whatever reason and the provided ROM is actually now loaded and decompressed if a decompressed ROM isn't already in the directory.

fenhl commented:
> Unless this has been resolved, can you point some of these out?

Here's an example (from <https://github.com/Cuphat/OoT-Randomizer/blob/2a5521493d3699cd9a32fe65f0167393e301458c/EntranceShuffle.py#L693-L696>):

```py
def shuffle_one_way_priority_entrances(worlds: list[World], world: World, one_way_priorities: dict[str, tuple[list[str], list[str]]],
                                       one_way_entrance_pools: dict[str, list[Entrance]], one_way_target_entrance_pools: dict[str, list[Entrance]],
                                       locations_to_ensure_reachable: Iterable[Location], complete_itempool: list[Item],
                                       retry_count: int = 2) -> list[tuple[Entrance, Entrance]]:
```

> As far as I can tell, everything here follows the standard for this, that standard being to line the subsequent lines up with the start of the statement on the original line (unless it's an if statement in which case you add more tabs so that it doesn't get confused for the code contained within the if). And viewing everything on Github, nothing got too close to the edge of my screen, and I just have a standard 1080p monitor

I don't think hanging indents are ever a good idea, despite what PEP 8 says. I'd consider them outdated in much the same way as the 80-character line length limit, which is also a PEP 8 rule we don't follow. Screen real estate isn't my concern here (horizontal scrolling is a thing). They're just much more tedious to work with since they're not always aligned to multiples of 4 spaces, and if you change the code to the left of the reference point (e.g. the function name in the example above), you have to reindent the following lines. Personally I also find them slightly harder to read because the code doesn't continue at an easily predictable position.
